### PR TITLE
Enables Ray Sharing and Separates Classes

### PR DESCRIPTION
**Pull request by owenmylotte**
Ray sharing implementation allows for decreased memory use and speedup in initialization and solve.

### CODE REVIEW

File: src/boundarySolver/physics/sublimation.hpp [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #include "finiteVolume/processes/pressureGradientScaling.hpp"
 #include "io/interval/interval.hpp"
 #include "radiation/radiation.hpp"
+#include "radiation/surfaceRadiation.hpp"
```
Comment (owenmylotte) [inline src/boundarySolver/physics/sublimation.hpp:11]: This isn't needed.

File: src/radiation/radiation.hpp [hunk 2]
```diff
@@ -47,26 +47,43 @@ class Radiation : public utilities::Loggable<Radiation> {  //!< Cell solver prov
...
+    virtual void Initialize(const solver::Range& cellRange, ablate::domain::SubDomain& subDomain);
+
+    //    void Initialize1D(const solver::Range &cellRange);
+    static PetscReal ReallySolveParallelPlates(PetscReal z);
+    static PetscReal EInteg(int order, double x);
+    static PetscReal CSimp(PetscReal a, PetscReal b, std::vector<double>& f);
+    static PetscReal Temperature(PetscReal x);
 
     /** Get the subdomain */
+    //    void Register(std::shared_ptr<ablate::domain::SubDomain>);
```
Comment (owenmylotte) [inline src/radiation/radiation.hpp:70]: Delete unused code.

File: src/radiation/radiation.cpp [hunk 3]
```diff
@@ -557,24 +394,34 @@ void ablate::radiation::Radiation::Solve(Vec solVec, ablate::domain::Field tempe
...
                 }
             }
         }
     }
 
+    // In a second loop, get carrier particles and set their values equal to the values of the carriers of the access identifiers
+    // This avoids solving any redundant rays, and also avoids changing the logic too much.
+    // Here we actually want to iterate through all particles.
+    for (PetscInt ipart = 0; ipart < npoints; ipart++) {
+        carrier[ipart].Ij = rays[Key(&access[ipart])].Ij;
```
Comment (owenmylotte) [inline src/radiation/radiation.cpp:416]: Get a single reference to the ray segment and don't cell the Key 3 times

File: src/radiation/radiation.cpp [hunk 4]
```diff
@@ -734,33 +568,18 @@ void ablate::radiation::Radiation::Solve(Vec solVec, ablate::domain::Field tempe
...
+        PetscInt index = GetLossCell(iCell, losses, solDm);  //!< Get the cell that the losses should be calculated with
         DMPlexPointLocalFieldRead(auxDm, index, temperatureField.id, auxArray, &temperature);
-        PetscReal losses = 4 * ablate::utilities::Constants::sbc * *temperature * *temperature * *temperature * *temperature;
-        if (surface) losses /= 2;  //!< If this is a surface then losses will only leave the hemisphere
+        losses *= 4 * ablate::utilities::Constants::sbc * *temperature * *temperature * *temperature * *temperature;
         if (log) {
             PetscReal centroid[3];
             DMPlexComputeCellGeometryFVM(cellDM, iCell, nullptr, centroid, nullptr) >> checkError;  //!< Reads the cell location from the current cell
-            printf("%f %f %f %f %f %f\n", centroid[0], centroid[1], centroid[2], origin[iCell].intensity, losses, *temperature);
+            printf("%f %f %f %f %f %f %f\n", centroid[0], centroid[1], centroid[2], origin[iCell].intensity, losses, 0.0, *temperature);
```
Comment (owenmylotte) [inline src/radiation/radiation.cpp:582]: Get rid of this

File: src/radiation/radiation.hpp [hunk 5]
```diff
@@ -47,26 +47,33 @@ class Radiation : public utilities::Loggable<Radiation> {  //!< Cell solver prov
     };
 
     std::map<PetscInt, Origin> origin;
+    std::map<std::string, bool> presence;
```
Comment (owenmylotte) [inline src/radiation/radiation.hpp:50]: Comment this

File: src/radiation/raySharingRadiation.cpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                 * When the search particle finds a ray segment to draw from in the new domain, it will need to travel to the location of the last cell in this segment.
+                 * The segment will likely not be complete yet, so the search particle will need to step without picking up cells until it reaches the end of this partition and reaches a new
+                 * partition The particle should exit the domain in this fashion, leaving a carrier particle with access to the other local ray segment, and no unique segment to be identified to
+                 * it. The carrier particle will have an identifier to control its travel, and an identifier to control where it draws information from. These identifiers in effect point to the
+                 * same ray.
+                 * If the particle is native to this domain, then we don't want to mess with it at all because it's at the beginning of building a ray.
+                 */
+                if (access[newpoint].origin != identifier[ipart].origin) {
+                    PetscReal centroid[3];
+                    PetscInt numPoints = static_cast<int>(rays[Key(&access[newpoint])].cells.size());
```
Comment (owenmylotte) [inline src/radiation/raySharingRadiation.cpp:91]: Cast to PetscInt

File: CMakeLists.txt [hunk 7]
```diff
@@ -38,7 +38,7 @@ if (CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
 endif ()
 
 # Create the library for ablate
-add_library(ablateLibrary SHARED)
```
Comment (mmcgurn) [inline CMakeLists.txt:41]: Would you mind moving these into their corresponding cmake files?

**Review by mmcgurn — CHANGES_REQUESTED** [review]

File: src/boundarySolver/physics/sublimation.cpp [hunk 8]
```diff
@@ -38,7 +37,7 @@ void ablate::boundarySolver::physics::Sublimation::Setup(ablate::boundarySolver:
     // Register an optional output function
     bSolver.RegisterFunction(SublimationOutputFunction,
                              this,
-                             {"conduction", "extraRad", "regressionMassFlux", "radiation"},
+                             {"conduction", "extraRad", "regressionMassFlux"},
```
[the remark this answers is not in the document]
Comment (owenmylotte) [reply src/boundarySolver/physics/sublimation.cpp]: I had accidentally written over my changes earlier. This is fixed now.

File: CMakeLists.txt (hunk 7 above)
Comment (owenmylotte) [reply CMakeLists.txt]: The main CMake file was messed up somehow. I've updated that. Thanks for pointing it out.

**Review by mmcgurn — APPROVED** [review]

**Pull request merged**